hs/src/Reach/Simulator.hs

```diff
@@ -0,0 +1,477 @@
+    DLE_TokenInit _at _dlarg -> return V_Null
+    DLE_CheckPay _at _slcxtframes dlarg _maybe_dlarg -> interp dlarg
+    DLE_Wait _at dltimearg -> case dltimearg of
+      Left dlarg -> do
+        ev <- interp dlarg
+        case ev of
+          V_UInt n -> incrNWtime n $ return V_Null
+          _ -> impossible "unexpected error"
+      Right dlarg -> do
+        ev <- interp dlarg
+        case ev of
+          V_UInt n -> incrNWsecs n $ return V_Null
+          _ -> impossible "unexpected error"
+    DLE_PartSet _at _slpart dlarg -> interp dlarg
+    DLE_MapRef _at dlmvar dlarg -> do
+      g <- globalGet
+      let linstate = e_linstate g
+      ev <- interp dlarg
+      case ev of
+        V_Address acc -> do
+          return $ flip (M.!) acc $ (M.!) linstate dlmvar
+        _ -> impossible "unexpected error"
+    DLE_MapSet _at dlmvar dlarg maybe_dlarg -> do
+      e <- globalGet
+      let linst = e_linstate e
+      ev <- interp dlarg
+      case ev of
+        V_Address acc -> case maybe_dlarg of
+          Nothing -> do
+            let m = M.delete acc ((M.!) linst dlmvar)
+            globalSet $ e {e_linstate = M.insert dlmvar m linst}
+            return V_Null
+          Just dlarg' -> do
+            ev' <- interp dlarg'
+            let m = M.insert acc ev' ((M.!) linst dlmvar)
+            globalSet $ e {e_linstate = M.insert dlmvar m linst}
+            return V_Null
+        _ -> impossible "unexpected error"
+    DLE_Remote _at _slcxtframes _dlarg _string _dlpayamnt _dlargs _dlwithbill -> undefined
+    DLE_TokenNew _at dltokennew -> do
+      supply <- interp (dtn_supply dltokennew)
+      case supply of
+        V_UInt supply' -> do
+          ledgerNewToken supply' $ return V_Null
+        _ -> impossible "expression interpreter"
+    DLE_TokenBurn _at dlarg1 dlarg2 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        (V_UInt tok, V_UInt burn_amt) -> do
+          updateLedger 0 tok (burn_amt-) $ return V_Null
+        _ -> impossible "expression interpreter"
+    DLE_TokenDestroy _at dlarg -> do
+      ev <- interp dlarg
+      case ev of
+        V_UInt tok -> do
+          e <- globalGet
+          let ledger = e_ledger e
+          let map_ledger = nw_ledger ledger
+          let new_nw_ledger = M.insert 0 (M.delete tok ((M.!) map_ledger 0)) map_ledger
+          globalSet $ e {e_ledger = ledger { nw_ledger = new_nw_ledger }}
+          return V_Null
+        _ -> impossible "expression interpreter"
+    -- TODO: new stuff
+    DLE_TimeOrder _at _assoc_maybe_arg_vars -> undefined
+    DLE_GetContract _at -> undefined
+    DLE_GetAddress _at -> undefined
+    DLE_EmitLog at _str dlvar -> interp $ DL_Var at dlvar
+    DLE_setApiDetails _at _slpart _dltpes _maybe_str -> undefined
+
+instance Interp DLStmt where
+  interp = \case
+    DL_Nop _at -> return V_Null
+    DL_Let _at let_var expr -> case let_var of
+      DLV_Eff -> do
+        _ <- interp expr
```

Comment:
I like `void $` for this pattern, even though it is more letters, I find it clearer

Reply:
i feel like most Haskell people would disagree :)

Reply:
based on the GHC warning and a quick reddit search